js/workspace_selection.js

```diff
@@ -0,0 +1,135 @@
+        inputElem.click((e) => e.stopPropagation()); // Stop click from closing
+        titleSpan.replaceWith(inputElem);
+        inputElem.val(workspaces[activeId].title)
+        inputElem.focus();
+        inputElem.on('focusout', () => {
+            inputElem.replaceWith(titleSpan);
+            editButton.show();
+            onSaveWorkspace(activeId, inputElem.val())
+        });
+
+        e.preventDefault();
+    }
+
+    function update() {
+        menuElem.empty();
+        // Sort for non-programmers with 1-9 and then 0 last. 
+        for (let index = 1; index < 11; index++) {
+            let id = index % 10
+            let element = workspaces[id % 10];
+
+            var item = $('<li></li>');
+            var link = $('<a href="#"></a>')
+
+            if (!element) {
+                // item.addClass("disabled");
+            }
+
+            var number = $('<span class="workspace-selector-index">').text(id);
+            var title = $('<span class="workspace-selector-title">')
+
+            if (!element) {
+                title.text("<empty>");
+                title.addClass("faded");
+            }
+            else {
+                title.text(element.title);
+            }
+
+            link.click((e) => {
+                if (element) {
+                    buttonElem.dropdown("toggle");
+                    onSelectionChange(workspaces, id);
+                    e.preventDefault();
+                }
+            });
+
+            var actionButtons = $('<span class="pull-right"></span>');
+
+            var saveButton = $('<span class="glyphicon glyphicon-floppy-disk" aria-hidden="true" data-toggle="tooltip" title="Save current graph setup to this Workspace"></span>');
+            saveButton.click((e) => {
+                if (!element) {
+                    onSaveWorkspace(id, "Unnamed");
+                }
+                else {
+                    onSaveWorkspace(id, element.title);
+                }
+                e.preventDefault();
+            });
+
+            saveButton.tooltip({ trigger: "hover", placement: "auto bottom" });
+
+            item.append(link);
+            link.append(number);
+            link.append(title);
+            link.append(actionButtons);
+            actionButtons.append(saveButton);
+            item.toggleClass("active", id == activeId)
+            menuElem.append(item);
+        }
+
+        if (workspaces[activeId]) {
+            numberSpan.text(activeId);
+            titleSpan.text(workspaces[activeId].title);
+        }
+        else {
+            titleSpan.text("");
+        }
+    }
+
+    this.setWorkspaces = function (newWorkspaces) {
+        workspaces = newWorkspaces;
+        update();
+    }
+
+    this.setActiveWorkspace = function (newId) {
+        activeId = newId;
+        update();
+    }
+
+    //TODO Should listen to workspace changes
+    // config.addListener(buildUI);
```

Comment:
This is a work in progress or this TODO is for a future reference?

Reply:
If we make it possible to edit the workspace title (or reorder them) from any other place, the changer will have to remember to trigger the update the control by calling .setWorkspaces or .setActiveWorkspace. So this is not needed now, but might be needed in the future. Maybe remove now and since it's hardy the only place in the code where we have this issue... :(